### PR TITLE
A decorator for Metrics that works without SimpleTraits

### PR DESCRIPTION
**Pull request by kellertuer**
As discussed in #48, I implemented the current decorator anew to cover the three cases we use more rigorously

* A manifold without a metric can still just be implemented
* A manifold with two different metrics that have closed forms can be easily implemented
* Adding a second metric to the case from the first point can easily be done, even with both being explicit `MetricManifolds` using the `is_default_metric`
* A metric with a `g` local tensor can be added to a manifold, with one restriction: It is not useful to set that one to default, since that would deactivate using the tensor again. Default should only be used for these implicit ones with closed formulae

The documentation is also already done, so I look forward to having feedback on this new model.

### CODE REVIEW

File: docs/src/manifolds/metric.md [hunk 1]
```diff
@@ -1,7 +1,46 @@
 # Metric manifold
 
+A Riemannian manifold always consists of a [topological manifold](https://en.wikipedia.org/wiki/Topological_manifold) together with a smoothly varying metric $g$.
+
+However, often there is a implicitly assumed (daefault) metric, like the usual
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]:
```suggestion
However, often there is an implicitly assumed (default) metric, like the usual
```

File: docs/src/manifolds/metric.md [hunk 2]
```diff
@@ -1,7 +1,46 @@
...
+inner product on [`Euclidean`](@ref) space. This decorator takes this into
+account. Hence it is not necessary to acts as a (as transparent as possible)
+decorator, see [`is_decorator_manifold`](@ref) for details.
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]: I don't really understand this sentence ("Hence it (...) for details"). Could you rephrase it?

File: docs/src/manifolds/metric.md [hunk 3]
```diff
@@ -1,7 +1,46 @@
...
+
+This manifold servers two purposes: To implement different metrics (i.g. in closed
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]:
```suggestion
This manifold decorator serves two purposes: to implement different metrics (e.g. in closed
```

File: docs/src/manifolds/metric.md [hunk 4]
```diff
@@ -1,7 +1,46 @@
...
+form) for one [`Manifold`](@ref) and to provide a way to compute geodesics
+on manifolds, where this [`Metric`](@ref) does not yield closed formula.
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]: I think eventually all other functions would also have a default implementation based on the metric, right? Like `exp`, `log`, parallel transport etc.

File: docs/src/manifolds/metric.md [hunk 5]
```diff
@@ -1,7 +1,46 @@
...
+
+## Implement Different Metrics on the same Manifold
+
+In order to distinguish different metrics on one manifold, one can introduce
+two [`Metric`](@ref)s and use this type to dispatch on the metric, see 
+[`SymmetricPositiveDefinite`](@ref). To avoid overhead, one [`Metric`](@ref)
+can then be marked as being the default, i.e. the one that is used, when no
+[`MetricManifold`](@ref) is used. This avoids reimplementation of a first
+existing metric, access for the default metric functions (without an [`MetricManifold`](@ref)
+overhead) as well as the transparent fallback of the corresponding [`MetricManifold`](@ref)
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]: Just to clarify, `MetricManifold` used in this way should not cause any runtime overhead.

File: docs/src/manifolds/metric.md [hunk 6]
```diff
@@ -1,7 +1,46 @@
...
+account. Hence it is not necessary to acts as a (as transparent as possible)
+decorator, see [`is_decorator_manifold`](@ref) for details.
+
+This manifold servers two purposes: To implement different metrics (i.g. in closed
+form) for one [`Manifold`](@ref) and to provide a way to compute geodesics
+on manifolds, where this [`Metric`](@ref) does not yield closed formula. 
+
+Let's first look at the provided types.
+
+## types
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]:
```suggestion
## Types
```

File: docs/src/manifolds/metric.md [hunk 7]
```diff
@@ -1,7 +1,46 @@
...
+In order to distinguish different metrics on one manifold, one can introduce
+two [`Metric`](@ref)s and use this type to dispatch on the metric, see 
+[`SymmetricPositiveDefinite`](@ref). To avoid overhead, one [`Metric`](@ref)
+can then be marked as being the default, i.e. the one that is used, when no
+[`MetricManifold`](@ref) is used. This avoids reimplementation of a first
+existing metric, access for the default metric functions (without an [`MetricManifold`](@ref)
+overhead) as well as the transparent fallback of the corresponding [`MetricManifold`](@ref)
+with default metric to the (classical) implementations.
+
+## Implementaion of Metrics
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]:
```suggestion
## Implementation of Metrics
```

File: docs/src/manifolds/metric.md [hunk 8]
```diff
@@ -1,7 +1,46 @@
...
+## Implement Different Metrics on the same Manifold
+
+In order to distinguish different metrics on one manifold, one can introduce
+two [`Metric`](@ref)s and use this type to dispatch on the metric, see 
+[`SymmetricPositiveDefinite`](@ref). To avoid overhead, one [`Metric`](@ref)
+can then be marked as being the default, i.e. the one that is used, when no
+[`MetricManifold`](@ref) is used. This avoids reimplementation of a first
+existing metric, access for the default metric functions (without an [`MetricManifold`](@ref)
+overhead) as well as the transparent fallback of the corresponding [`MetricManifold`](@ref)
+with default metric to the (classical) implementations.
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]: Maybe "undecorated" would be clearer than "classical"?

File: docs/src/manifolds/metric.md [hunk 9]
```diff
@@ -1,7 +1,46 @@
...
+Order = [:type]
+```
+
+## Implement Different Metrics on the same Manifold
+
+In order to distinguish different metrics on one manifold, one can introduce
+two [`Metric`](@ref)s and use this type to dispatch on the metric, see 
+[`SymmetricPositiveDefinite`](@ref). To avoid overhead, one [`Metric`](@ref)
+can then be marked as being the default, i.e. the one that is used, when no
+[`MetricManifold`](@ref) is used. This avoids reimplementation of a first
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]:
```suggestion
[`MetricManifold`](@ref) is used. This avoids reimplementation of the first
```

File: docs/src/manifolds/metric.md [hunk 10]
```diff
@@ -1,7 +1,46 @@
...
+existing metric, access for the default metric functions (without an [`MetricManifold`](@ref)
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]:
```suggestion
existing metric, access to the default metric functions (without an [`MetricManifold`](@ref)
```

Perhaps also "default metric functions" should be replaced by a more precise phrase, maybe "metric-dependent functions of the undecorated manifold"?

File: src/Metric.jl [hunk 11]
```diff
@@ -27,23 +27,55 @@ abstract type LorentzMetric <: Metric end
...
+This means that any occurence of
+[`MetricManifold`](@ref)`(M,G)` where `is_default_metric(M,G) = Val{true}`
+falls back to just be called with `M` such that the [`Manifold`](@ref) `M`
+implicitly has this metric, for example if this was the first one implemented
+or is the one most commonly assumed to be used.
+"""
+is_default_metric(M::MMT) where {MMT <: MetricManifold} = is_default_metric(base_manifold(M),metric(M))
+# this automatically undecorates
+convert(::Type{MT},M::MetricManifold{<:MT,GT}) where {MT <: Manifold,GT} = base_manifold(M)
+# this should austomatically decorate at least for simple cases
```
Comment (mateuszbaran) [inline src/Metric.jl]:
```suggestion
# this should automatically decorate at least for simple cases
```

**Comment by mateuszbaran** [comment]
Probably the SPD manifold needs some changes since coverage says that some methods are not called, I guess the new metric manifold implementation makes some methods redundant?

File: docs/src/manifolds/metric.md (hunk 4 above)
Comment (kellertuer) [reply docs/src/manifolds/metric.md]: Well I hope so, yes, for now there is only exp implemented based on the metric tensor. I have not yet read how to do the remaining ones numerically – but also since my goal is more to implement different metrics

File: docs/src/manifolds/metric.md (hunk 5 above)
Comment (kellertuer) [reply docs/src/manifolds/metric.md]: Yes, it should not cause overhead. I'll add that.

**Comment by kellertuer** [comment]
Interestingly I do a for loop over metric manifolds and (added a print temporarily) it seems to take always the first of the for loop entries, that's why SPD never checks LogCholesky. Have to look closer into that. However, Metrics should be covered better now.

**Comment by kellertuer** [comment]
Just had a moment of time and finished the final remarks as well as took some time to increase test coverage (still a special case in rotations missing for example), so despite (slowly) increasing test coverage(even more); this is ready to merge.

**Comment by kellertuer** [comment]
I'll wait a few days for everyone to check, but from implementations and testing as well as documentation I think this is finished.

File: docs/src/manifolds/metric.md [hunk 12]
```diff
@@ -1,7 +1,51 @@
 # Metric manifold
 
+A Riemannian manifold always consists of a [topological manifold](https://en.wikipedia.org/wiki/Topological_manifold) together with a smoothly varying metric $g$.
+
+However, often there is an implicitly assumed (default) metric, like the usual
+inner product on [`Euclidean`](@ref) space. This decorator takes this into
+account. It is not necessary to use this decorator, if you imlement just one (or
+the first) metric. If you later introduce a second, the old (first) one can be
+used with or without a metricx. The decorator acts transparent in that sense;
```
Comment (mateuszbaran) [inline docs/src/manifolds/metric.md]:
```suggestion
used with or without a metric. The decorator acts transparent in that sense;
```

File: src/Metric.jl [hunk 13]
```diff
@@ -27,23 +27,54 @@ abstract type LorentzMetric <: Metric end
...
+[`MetricManifold`](@ref)(M,G) where `is_default_metric(M,G) = Val{true}`
+falls back to just be called with `M` such that the [`Manifold`](@ref) `M`
+implicitly has this metric, for example if this was the first one implemented
+or is the one most commonly assumed to be used.
+"""
+is_default_metric(M::Manifold,G::Metric) = Val(false)
+"""
+    is_default_metric(MM)
+
+indicate, whether the [`Metric`](@ref) `G` is the default metric for
```
Comment (mateuszbaran) [inline src/Metric.jl]: I think in this docstring `MM.G` and `MM.M` should be used instead of just `G` and `M` (because the function only takes one argument called `MM`).

**Comment by mateuszbaran** [comment]
Apart from these two docstring issues I think it can be merged 👍

**Comment by sethaxen** [comment]
It's a holiday weekend here. I should be able to review by Tuesday.

File: src/Metric.jl (hunk 13 above)
Comment (kellertuer) [reply src/Metric.jl]: Thanks for carefully reading even the docstrings; you're right that makes the statements far clearer.

**Comment by kellertuer** [comment]
Nevermind,if it gets Tuesday. I just had a (long) train ride yesterday and one (not so long) today (and one tomorrow...) so I finished this from my side (when I got too lazy to write the remaining few tests). Feel free to take your time to review this; but I would like to make sure, that you're ideas are still preserved as you intended them :)

File: src/Metric.jl [hunk 14]
```diff
@@ -27,23 +27,54 @@ abstract type LorentzMetric <: Metric end
...
-[`local_metric`](@ref).
+Equip a [`Manifold`](@ref) explicitly with a [`Metric`](@ref) `G`. 
 
-# Constructor
+For a Metric Manifold, by default, assumes, that you implement the linear form
+from [`local_metric`](@ref) in order to evaluate the exponential map.
+
+If the corresponding [`Metric`](@ref) `G` yields closed form formulae for e.g.
+the exponential map and this is implemented directly (without solving the ode),
+you of course can still implement that directly.
```
[the remark this answers is not in the document]
Comment (kellertuer) [reply src/Metric.jl]: Thanks for the grammar correction, I'll remove the "can" after "of course" then.

File: docs/src/manifolds/metric.md [hunk 15]
```diff
@@ -1,7 +1,51 @@
 # Metric manifold
 
+A Riemannian manifold always consists of a [topological manifold](https://en.wikipedia.org/wiki/Topological_manifold) together with a smoothly varying metric $g$.
+
+However, often there is an implicitly assumed (default) metric, like the usual
+inner product on [`Euclidean`](@ref) space. This decorator takes this into
+account. It is not necessary to use this decorator, if you imlement just one (or
+the first) metric. If you later introduce a second, the old (first) one can be
```
[the remark this answers is not in the document]
Comment (kellertuer) [reply docs/src/manifolds/metric.md]: I avoided the second "one" by rephrasing the sentence.
